connectors/sources/google_drive.py

```diff
@@ -0,0 +1,638 @@
+                        api_name=API_NAME, api_version=API_VERSION
+                    )
+                    if RUNNING_FTEST and not sub_method and GOOGLE_DRIVE_EMULATOR_HOST:
+                        logger.debug(
+                            f"Using the google drive emulator at {GOOGLE_DRIVE_EMULATOR_HOST}"
+                        )
+                        # Redirecting calls to fake Google Drive server for e2e test.
+                        drive_client.discovery_document["rootUrl"] = (
+                            GOOGLE_DRIVE_EMULATOR_HOST + "/"
+                        )
+                    resource_object = getattr(drive_client, resource)
+                    method_object = getattr(resource_object, method)
+                    if full_response:
+                        first_page_with_next_attached = (
+                            await google_client.as_service_account(
+                                method_object(**kwargs),
+                                full_res=True,
+                                timeout=DRIVE_API_TIMEOUT,
+                            )
+                        )
+
+                        async for page_items in first_page_with_next_attached:
+                            yield page_items
+                            retry_counter = 0
+                    else:
+                        if sub_method:
+                            method_object = getattr(method_object, sub_method)
+                        yield await google_client.as_service_account(
+                            method_object(**kwargs), timeout=DRIVE_API_TIMEOUT
+                        )
+                    break
+            except AttributeError as exception:
+                logger.error(
+                    f"Error occurred while generating the resource/method object for an API call. Error: {exception}"
+                )
+                raise exception
+            except HTTPError as exception:
+                log_message = f"Retry count: {retry_counter} out of {self.retry_count}. Response code: {exception.res.status_code} Exception: {exception}."
+                retry_counter = await self._retry(retry_counter, log_message)
+            except Exception as exception:
+                log_message = f"Retry count: {retry_counter} out of {self.retry_count}. Exception: {exception}."
+                retry_counter = await self._retry(retry_counter, log_message)
+
+    async def _retry(self, retry_counter, log_message):
+        """Retry logic for handling Google API calls
+
+        Args:
+            retry_counter (int): Current number of attempted retries
+            log_message (str): Warning message logged
+
+        Returns:
+            int: increased retry_counter
+        """
+        retry_counter += 1
+        if retry_counter > self.retry_count:
+            raise
+        logger.warning(log_message)
+        await asyncio.sleep(DEFAULT_WAIT_MULTIPLIER**retry_counter)
```

Comment:
Avoid using `asyncio.sleep` and use `CancellableSleeps` from `utils.py`.

Problem with `asyncio.sleep` is that it will keep the process running if you want to terminate it, for example you do `await asyncio.sleep(60)` and then CTRL+C the host - it'll have to wait until `asyncio.sleep` wakes up, which is 60 seconds - too much. We want to be able to cancel these sleeps as well.